common/io/io.cpp

```diff
@@ -92,50 +92,93 @@ Error Copy(Writer &dst, Reader &src, vector<uint8_t> &buffer) {
 	}
 }
 
-void AsyncCopy(Writer &dst, AsyncReader &src, function<void(Error)> finished_handler) {
+struct CopyData {
+	CopyData(size_t limit) :
+		buf(MENDER_BUFSIZE),
+		limit(limit) {
+	}
+
+	vector<uint8_t> buf;
+	size_t copied {0};
+	size_t limit;
+};
+
+void AsyncCopy(
+	Writer &dst, AsyncReader &src, function<void(Error)> finished_handler, size_t stop_after) {
 	AsyncCopy(
 		WriterPtr(&dst, [](Writer *) {}),
 		AsyncReaderPtr(&src, [](AsyncReader *) {}),
 		finished_handler);
 }
 
-void AsyncCopy(WriterPtr dst, AsyncReaderPtr src, function<void(Error)> finished_handler) {
-	auto buf = make_shared<vector<uint8_t>>(MENDER_BUFSIZE);
-	src->RepeatedAsyncRead(
-		buf->begin(), buf->end(), [dst, src, buf, finished_handler](ExpectedSize size) {
+void AsyncCopy(
+	WriterPtr dst, AsyncReaderPtr src, function<void(Error)> finished_handler, size_t stop_after) {
+	auto data = make_shared<CopyData>(stop_after);
+	class Functor {
+	public:
+		void operator()(ExpectedSize size) {
 			if (!size) {
 				finished_handler(size.error());
-				return Repeat::No;
+				return;
 			}
 
+
 			if (*size == 0) {
 				finished_handler(error::NoError);
-				return Repeat::No;
+				return;
 			}
 
-			auto n = dst->Write(buf->begin(), buf->begin() + *size);
+			auto n = writer->Write(data->buf.begin(), data->buf.begin() + *size);
 			if (!n) {
 				finished_handler(n.error());
-				return Repeat::No;
+				return;
 			} else if (*n != *size) {
 				finished_handler(Error(
 					make_error_condition(std::errc::io_error), "Short write when copying data"));
-				return Repeat::No;
+				return;
 			}
 
-			return Repeat::Yes;
-		});
+			data->copied += *n;
+
+			size_t to_copy = min(data->limit - data->copied, data->buf.size());
+			if (to_copy == 0) {
+				finished_handler(error::NoError);
+				return;
+			}
+
+			auto err = reader->AsyncRead(
+				data->buf.begin(),
+				data->buf.begin() + to_copy,
+				Functor {writer, reader, data, finished_handler});
+			if (err != error::NoError) {
+				finished_handler(err);
+			}
+		}
+
+		WriterPtr writer;
+		AsyncReaderPtr reader;
+		shared_ptr<CopyData> data;
+		function<void(Error)> finished_handler;
+	};
+	size_t to_copy = min(data->limit, data->buf.size());
+	auto err = src->AsyncRead(
+		data->buf.begin(), data->buf.begin() + to_copy, Functor {dst, src, data, finished_handler});
+	if (err != error::NoError) {
+		finished_handler(err);
+	}
 }
 
-void AsyncCopy(AsyncWriter &dst, Reader &src, function<void(Error)> finished_handler) {
+void AsyncCopy(
+	AsyncWriter &dst, Reader &src, function<void(Error)> finished_handler, size_t stop_after) {
 	AsyncCopy(
 		AsyncWriterPtr(&dst, [](AsyncWriter *) {}),
 		ReaderPtr(&src, [](Reader *) {}),
 		finished_handler);
```

Comment:
Same here, missing to pass `stop_after`. It is also missing in the rest of versions of `AsyncCopy` below.